src/localplayer.cpp

```diff
@@ -1077,3 +1074,82 @@ float LocalPlayer::getSlipFactor(Environment *env, const v3f &speedH)
 	}
 	return 1.0f;
 }
+
+void LocalPlayer::handleAutojump(f32 dtime, Environment *env,
+		const collisionMoveResult &result, const v3f &position_before_move,
+		const v3f &speed_before_move, f32 pos_max_d)
+{
+	PlayerSettings &player_settings = getPlayerSettings();
+	if (!player_settings.autojump)
+		return;
+
+	if (m_autojump) {
+		// release autojump after a given time
+		m_autojump_time -= dtime;
+		if (m_autojump_time <= 0.0f)
+			m_autojump = false;
+		return;
+	}
+
+	bool control_forward = control.up ||
+			       (!control.up && !control.down &&
+					       control.forw_move_joystick_axis < -0.05);
+	bool could_autojump =
+			m_can_jump && !control.jump && !control.sneak && control_forward;
+	if (!could_autojump)
+		return;
+
+	bool horizontal_collision = false;
+	for (const auto &colinfo : result.collisions) {
+		if (colinfo.type == COLLISION_NODE && colinfo.plane != 1) {
+			horizontal_collision = true;
+			break; // one is enough
+		}
+	}
+
+	// must be running against something to trigger autojumping
+	if (!horizontal_collision)
+		return;
+
+	// check for nodes above
+	v3f headpos_min = m_position + m_collisionbox.MinEdge * 0.99f;
+	v3f headpos_max = m_position + m_collisionbox.MaxEdge * 0.99f;
+	headpos_min.Y = headpos_max.Y; // top face of collision box
+	v3s16 ceilpos_min = floatToInt(headpos_min, BS) + v3s16(0, 1, 0);
+	v3s16 ceilpos_max = floatToInt(headpos_max, BS) + v3s16(0, 1, 0);
+	const NodeDefManager *ndef = env->getGameDef()->ndef();
+	bool is_position_valid;
+	for (s16 z = ceilpos_min.Z; z <= ceilpos_max.Z; z++) {
+		for (s16 x = ceilpos_min.X; x <= ceilpos_max.X; x++) {
+			MapNode n = env->getMap().getNodeNoEx(v3s16(x, ceilpos_max.Y, z), &is_position_valid);
+
+			if (!is_position_valid)
+				break;  // won't collide with the void outside
+			if (n.getContent() == CONTENT_IGNORE)
+				return; // players collide with ignore blocks -> same as walkable
+			const ContentFeatures &f = ndef->get(n);
+			if (f.walkable)
+				return; // would bump head, don't jump
+		}
+	}
+
+	float jump_height = 1.1f; // TODO: better than a magic number
+	v3f jump_pos = position_before_move + v3f(0.0f, jump_height * BS, 0.0f);
+	v3f jump_speed = speed_before_move;
+
+	// try at peak of jump, zero step height
+	collisionMoveResult jump_result = collisionMoveSimple(env, m_client, pos_max_d,
+			m_collisionbox, 0.0f, dtime, &jump_pos, &jump_speed,
+			v3f(0, 0, 0));
+
+	// see if we can get a little bit farther horizontally if we had
+	// jumped
+	v3f run_delta = position_before_move - m_position;
+	run_delta.Y = 0.0f;
+	v3f jump_delta = position_before_move - jump_pos;
```

Comment:
Here and above. Mathematically works but the order of terms bothers me because a movement vector is 'final pos' - 'start pos'.